provider/openstack/firewaller.go

```diff
@@ -70,16 +71,341 @@ type firewallerFactory struct {
 
 // GetFirewaller implements FirewallerFactory
 func (f *firewallerFactory) GetFirewaller(env environs.Environ) Firewaller {
-	return &defaultFirewaller{environ: env.(*Environ)}
+	return &switchingFirewaller{env: env.(*Environ)}
+}
+
+type switchingFirewaller struct {
+	env *Environ
+
+	mu sync.Mutex
+	fw Firewaller
+}
+
+func (f *switchingFirewaller) initFirewaller() error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if f.fw != nil {
+		return nil
+	}
+
+	client := f.env.client()
+	if !client.IsAuthenticated() {
+		if err := authenticateClient(client); err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	base := firewallerBase{environ: f.env}
+	if f.env.supportsNeutron() {
+		f.fw = &neutronFirewaller{base}
+	} else {
+		f.fw = &legacyNovaFirewaller{base}
+	}
+	return nil
+}
+
+func (f *switchingFirewaller) OpenPorts(ports []network.PortRange) error {
+	if err := f.initFirewaller(); err != nil {
+		return errors.Trace(err)
+	}
+	return f.fw.OpenPorts(ports)
+}
+
+func (f *switchingFirewaller) ClosePorts(ports []network.PortRange) error {
+	if err := f.initFirewaller(); err != nil {
```

Comment:
Lazy auth is to avoid API calls in EnvironProvider.Open, which we call very frequently.

One option is to defer construction to the point where we make the calls, but that means that constructors all need to be cheap.

I'm going to defer this for now, in the hope that we can drop the switching impls soon. If not, I'll revisit.